spec/Appendix A -- Persisted Documents.md

```diff
@@ -95,6 +95,54 @@ implementation specific.
 Note: A 32 character hexadecimal _custom document identifier_ is likely to be an
 MD5 hash of the GraphQL document, as traditionally used by Relay.
 
+### Document identifier syntax
+
+A _document identifier_ must only contain colons (`:`) and characters that are
+defined as
+[`unreserved` in RFC3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3)
+(alphanumeric characters (`A-Z`, `a-z`, `0-9`), dashes (`-`), periods (`.`),
+underscores (`_`), and tildes (`~`)).
```

Comment:
Can we add `/` to the list of allowed characters?

For GraphQL Hive, we already use `"<app_name>/<app_version>/<document_hash>"` as the document id for namespacing documents. (https://the-guild.dev/graphql/hive/docs/features/app-deployments#sending-persisted-document-requests-from-your-app)

From a persisted document store point of view, it does not make sense for us to share the hashes between all the different GraphQL clients.

You might not own all the teams that write to the persisted document store and dictating the hash algorithm to be used for everyone might not make sense.

In addition, we use the extracted app_name and app_version for analytics of the GraphQL API client usage, but that is irrelevant for this request, but might be relevant for the motivations.

Reply:
That'd works for me 👍 . 

IIRC, there was a desire that the identifier should be "url-compatible" to use it as part of the url with GET request and improve the debugging/monitoring experience. And `/` might have a specific meaning there. I'm personally fine with letting implementers deal with that.

I'll ultimately defer to @benjie on this one as he has put a lot of thought into this.

Reply:
Already part of the spec:

https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#get

> For HTTP GET requests, the GraphQL-over-HTTP request parameters MUST be provided in the query component of the request URL, encoded in the application/x-www-form-urlencoded format as specified by the [WhatWG URLSearchParams class](https://url.spec.whatwg.org/#interface-urlsearchparams).

Reply:
To be clear, the GET requests being recommended as part of #305 are `/graphql/<hash>/<operationname>`, not using query strings. Technically you can have `hash` here have `%2F` in it to escape the `/`, but I think it would be better to stick to URL-safe characters if possible.

Happy to change a `MUST` to a `SHOULD` such that `/` is allowed. E.g. we RECOMMEND that you only use characters in the given set, but you MAY use others if there's a good reason to do so.

Before doing so, though, I'd ask whether using `~` might be a more acceptable character substitute - suitably uncommon, but also URL-safe.

Reply:
> Happy to change a `MUST` to a `SHOULD` such that `/` is allowed

Personally, I'd do this, along with a note about encoding the hash alongside the `/graphql/<hash>/<operationname>` recommendation in #305 .  I probably would not add `/` to the safe list of characters when we are simultaneously recommending `/` be used to separate the hash from the operation name.

So the additional note might read something like this:
> The document id / hash MUST be encoded using percent-encoding as specified by the [WhatWG URLSearchParams class](https://url.spec.whatwg.org/#interface-urlsearchparams).

We should not need to describe how to encode the document id as a query parameter (e.g. `/graphql?documentId=<hash>`) as this is already specified in the spec.

Reply:
This made me aware of something else, doesnt `/graphql/<sha>` already conflict with /graphql/stream used by the GraphQL over SSE protocol?

Would in that case it make sense to use a different pattern such as `/graphql/persisted/sha`?

That would still not solve the /operation-name appendix though..

Using tilde is okay, but certainly looks weird.

Reply:
I've been looking at the GraphQL over SSE specification.  I can't find that the url is defined anywhere, besides being used in some samples - ??  Seems to me that when the server receives `Accept: text/event-stream` then it may (not must, in case of validation errors) respond with a SSE stream...

Reply:
@Shane32 You are right it is not part of the specification. But the reference implementation `graphql-sse` kind of promotes using `/graphql/stream`: https://the-guild.dev/graphql/sse/recipes

It is out of topic here though, and I am fine with changing `/` to `~`. It might be worth to also ask other vendors that provide a persisted query feature though, since it can imply breaking changes.

Reply:
PostGraphile also uses `/graphql/stream` as the endpoint to subscribe to schema change notifications as specified in https://github.com/graphql/graphql-over-http/issues/48

I agree with using a prefix.